src/util/RtreeBuild.cpp

```diff
@@ -0,0 +1,1133 @@
+    }
+
+    if (ySize == 0) {
+      minD1 = element;
+      maxD1 = element;
+    }
+    if (element.orderY > maxD1.orderY) {
+      maxD1 = element;
+    }
+
+    ySize++;
+  };
+
+  if (workInRam) {
+    for (RTreeValueWithOrderIndex element : *RectanglesD1WithOrder) {
+      processD1Element(element);
+
+      RectanglesD0WithOrder->push_back(element);
+    }
+    centerOrdering(*RectanglesD0WithOrder, 0);
+  } else {
+    for (RTreeValueWithOrderIndex element : sorterRectsD1.sortedView()) {
+      processD1Element(element);
+
+      Rtree::SaveEntryWithOrderIndex(element, r1File);
+      sorterRectsD0.push(element);
+    }
+  }
+
+  r1File.close();
+  sorterRectsD1.clear();
+
+  // replace the placeholder
+  (*r1Small)[0] = minD1;
+  (*r1Small)[1] = maxD1;
+
+  uint64_t currentX = 0;
+  std::ofstream r0File =
+      std::ofstream(onDiskBase + ".boundingbox.d0.tmp", std::ios::binary);
+  std::shared_ptr<multiBoxWithOrderIndex> r0Small =
+      std::make_shared<multiBoxWithOrderIndex>();
+  // placeholder
+  r0Small->push_back(RTreeValueWithOrderIndex());
+  r0Small->push_back(RTreeValueWithOrderIndex());
+  RTreeValueWithOrderIndex minD0;
+  RTreeValueWithOrderIndex maxD0;
+
+  auto processD0Element = [&currentX, currentS, M, &r0Small, &minD0,
+                           &maxD0](RTreeValueWithOrderIndex element) {
+    if (((currentX + 1) % currentS == 0 && (currentX + 1) / currentS >= 1 &&
+         (currentX + 1) / currentS < M) ||
+        (currentX % currentS == 0 && currentX / currentS >= 1 &&
+         currentX / currentS < M)) {
+      // index i * S - 1 or i * S
+      r0Small->push_back(element);
+    }
+
+    if (currentX == 0) {
+      minD0 = element;
+      maxD0 = element;
+    }
+    if (element.orderX > maxD0.orderX) {
+      maxD0 = element;
+    }
+
+    currentX++;
+  };
+
+  if (workInRam) {
+    for (RTreeValueWithOrderIndex element : *RectanglesD0WithOrder) {
+      processD0Element(element);
+    }
+  } else {
+    for (RTreeValueWithOrderIndex element : sorterRectsD0.sortedView()) {
+      Rtree::SaveEntryWithOrderIndex(element, r0File);
+
+      processD0Element(element);
+    }
+  }
+
+  r0File.close();
+  sorterRectsD0.clear();
+
+  // replace the placeholder
+  (*r0Small)[0] = minD0;
+  (*r0Small)[1] = maxD0;
+
+  RectanglesForOrderedBoxes rectsD0;
+  RectanglesForOrderedBoxes rectsD1;
+  rectsD0.rectanglesSmall = r0Small;
+  rectsD1.rectanglesSmall = r1Small;
+  if (workInRam) {
+    rectsD0.rectanglesInRam = RectanglesD0WithOrder;
+    rectsD1.rectanglesInRam = RectanglesD1WithOrder;
+    orderedInputRectangles.SetOrderedBoxesToRam(rectsD0, rectsD1, boundingBox);
+  } else {
+    rectsD0.rectanglesOnDisk = onDiskBase + ".boundingbox.d0";
+    rectsD1.rectanglesOnDisk = onDiskBase + ".boundingbox.d1";
+    orderedInputRectangles.SetOrderedBoxesToDisk(rectsD0, rectsD1, xSize,
+                                                 boundingBox);
+  }
+  return orderedInputRectangles;
+}
+
```

Comment:
The whole initial sorting can be split into its own class, because it is very complicated and takes quite a lot of code, and it is only done once in the beginning.